### PR TITLE
PYIC-7746: Embedded metrics with powertools

### PR DESCRIPTION
**Pull request by Wynndow**
### What changed

Embedded metrics with powertools

### Why did it change

Powertools have recently added support for sending embedded metrics. This uses that to send CRI redirect and return metrics, to allow us to track completion rates.

This will automatically add things like x-ray trace ids, and lambda request ids to the generated logs. We can also get cold start metrics for free.

Screenshot 2024-12-13 at 12 49 27

### Issue tracking

- [PYIC-7746](https://govukverify.atlassian.net/browse/PYIC-7746)

[PYIC-7746]: https://govukverify.atlassian.net/browse/PYIC-7746?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deploy/template.yaml [hunk 1]
```diff
@@ -51,6 +52,7 @@ Globals:
         DT_OPEN_TELEMETRY_ENABLE_INTEGRATION: "true"
         OTEL_INSTRUMENTATION_AWS_SDK_EXPERIMENTAL_USE_PROPAGATOR_FOR_MESSAGING: true
         JAVA_TOOL_OPTIONS: -XX:+TieredCompilation -XX:TieredStopAtLevel=1
+        POWERTOOLS_METRICS_NAMESPACE: !Sub CoreBackEmbeddedMetrics-${Environment}
```
Comment (Wynndow) [inline deploy/template.yaml:55]: Adding the environment suffix here is useful for separating dev-env metrics.

File: lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/ProcessCriCallbackHandler.java [hunk 2]
```diff
@@ -143,26 +143,9 @@ public ProcessCriCallbackHandler() {
     @Override
     @Tracing
     @Logging(clearState = true)
+    @Metrics(captureColdStart = true)
     public APIGatewayProxyResponseEvent handleRequest(
             APIGatewayProxyRequestEvent input, Context context) {
-        // temporary logging to check which tracing headers are being used by dynatrace
-        if ("build".equals(configService.getEnvironmentVariable(EnvironmentVariable.ENVIRONMENT))) {
```
Comment (Wynndow) [inline lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/ProcessCriCallbackHandler.java:149]: Temporary logging from an old PR which should have been removed. Naughty.

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/IpvSessionItem.java [hunk 3]
```diff
@@ -42,7 +42,7 @@ public class IpvSessionItem implements PersistenceItem {
     private long ttl;
     private String emailAddress;
     private ReverificationStatus reverificationStatus;
-    private List<String> stateStack = new ArrayList<>();
+    @Builder.Default private List<String> stateStack = new ArrayList<>();
```
Comment (Wynndow) [inline libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/IpvSessionItem.java:45]: This is fixing a compilation warning - lombok didn't like the mix of having a builder and having the field already initialised.

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java [hunk 4]
```diff
@@ -148,6 +150,7 @@ public BuildCriOauthRequestHandler(ConfigService configService) {
     @Override
     @Tracing
     @Logging(clearState = true)
+    @Metrics(captureColdStart = true)
```
Comment (Joe-Edwards-GDS) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java:153]: Does this capture cold start times in a metric? (Should we think about adding this to all our lambdas?)

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+import static uk.gov.di.ipv.core.library.helpers.EmbeddedMetricHelper.Dimension.CRI;
+import static uk.gov.di.ipv.core.library.helpers.EmbeddedMetricHelper.Metric.CRI_REDIRECT;
+import static uk.gov.di.ipv.core.library.helpers.EmbeddedMetricHelper.Metric.CRI_RETURN;
+import static uk.gov.di.ipv.core.library.helpers.LogHelper.GOVUK_SIGNIN_JOURNEY_ID_DEFAULT_VALUE;
+import static uk.gov.di.ipv.core.library.helpers.LogHelper.LogField.LOG_GOVUK_SIGNIN_JOURNEY_ID;
+
+@ExcludeFromGeneratedCoverageReport
+public class EmbeddedMetricHelper {
+    private static final MetricsLogger METRICS_LOGGER = MetricsUtils.metricsLogger();
```
Comment (Joe-Edwards-GDS) [inline libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java:19]: I assume it's fine to keep this as a static variable - i.e. it's safe to reload across snapstart?

We assign properties and dimensions directly to the logger, do they get reset between invocations?

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    private static void createMetrics(
+            Map<Dimension, String> dimensions,
+            Map<Metric, Double> metrics,
+            String govukSigninJourneyId) {
+        METRICS_LOGGER.putProperty(
+                LOG_GOVUK_SIGNIN_JOURNEY_ID.getFieldName(),
+                govukSigninJourneyId == null
+                        ? GOVUK_SIGNIN_JOURNEY_ID_DEFAULT_VALUE
+                        : govukSigninJourneyId);
```
Comment (Joe-Edwards-GDS) [inline libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java]: (Same comment as the other review about Log4J context)

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/IpvSessionItem.java (hunk 3 above)
Comment (Joe-Edwards-GDS) [reply libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/IpvSessionItem.java:45]: (Fine - although we basically never actually use the builder)

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    public static void criRedirect(String cri, String govukSigninJourneyId) {
+        createMetrics(Map.of(CRI, cri), Map.of(CRI_REDIRECT, 1.0), govukSigninJourneyId);
+    }
+
+    public static void criReturn(String cri, String govukSigninJourneyId) {
+        createMetrics(Map.of(CRI, cri), Map.of(CRI_RETURN, 1.0), govukSigninJourneyId);
+    }
+
+    private static void createMetrics(
```
Comment (Joe-Edwards-GDS) [inline libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java]: (Same comment as other review)

**Review by Joe-Edwards-GDS — COMMENTED** [review]
Not much in it, but this seems slightly better and avoids needing to worry about flushing, what do you think?

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/IpvSessionItem.java (hunk 3 above)
Comment (Wynndow) [reply libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/IpvSessionItem.java:45]: I think I introduced in pretty recently which is why there probably isn't much adoption of it yet - it was useful for tests

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java (hunk 4 above)
Comment (Wynndow) [reply lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java:153]: Yeah, if we go down this route that's probably a good idea

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java (hunk 5 above)
Comment (Wynndow) [reply libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java:19]: Yeah, the values get reset by the flush that happend at the end of every lambda. The metrics look good after running multiple journeys.

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java (hunk 6 above)
Comment (Wynndow) [reply libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java]: Yeah - I think it should be easily do-able. And I'm not sure that's a large risk of a key-clash. I'll give it a go.

Comment (Wynndow) [reply libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java]: Screenshot 2024-12-17 at 10 02 05

Looks like it works pretty good.

**Review by Joe-Edwards-GDS — APPROVED** [review]

**Pull request merged**